sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -327,121 +329,320 @@ bool WinHttpTransport::VerifyCertificatesInChain(
+   * with. If it is, we accept the connection, if it is not, we abort the connection, closing the
+   * incoming request handle.
+   */
+  void WinHttpAction::OnHttpStatusOperation(
+      HINTERNET hInternet,
+      DWORD internetStatus,
+      LPVOID statusInformation,
+      DWORD statusInformationLength)
   {
-    DWORD bufferLength = sizeof(PCCERT_CONTEXT);
-    if (!WinHttpQueryOption(
-            hInternet,
-            WINHTTP_OPTION_SERVER_CERT_CONTEXT,
-            reinterpret_cast<void*>(serverCertificate.addressof()),
-            &bufferLength))
+    Log::Write(
+        Logger::Level::Informational,
+        "Status operation: " + std::to_string(internetStatus) + "("
+            + InternetStatusToString(internetStatus) + ")");
+    if (internetStatus == WINHTTP_CALLBACK_STATUS_SECURE_FAILURE)
+    {
+      Log::Write(Logger::Level::Error, "Security failure. :(");
+    }
+    else if (internetStatus == WINHTTP_CALLBACK_STATUS_REQUEST_ERROR)
     {
-      GetErrorAndThrow("Could not retrieve TLS server certificate.");
+      WINHTTP_ASYNC_RESULT* asyncResult = static_cast<WINHTTP_ASYNC_RESULT*>(statusInformation);
+      Log::Write(
+          Logger::Level::Error,
+          "Request error. " + GetErrorMessage(asyncResult->dwError) + " "
+              + std::to_string(asyncResult->dwResult));
+      CompleteActionWithError(asyncResult->dwResult, asyncResult->dwError);
+    }
+    else if (internetStatus == WINHTTP_CALLBACK_STATUS_SENDING_REQUEST)
+    {
+      // We will only set the Status callback if a root certificate has been set. There is no action
+      // which needs to be completed for this notification.
+      m_httpRequest->HandleExpectedTlsRootCertificates(hInternet);
+    }
+    else if (internetStatus == m_expectedStatus)
+    {
+      if (internetStatus == WINHTTP_CALLBACK_STATUS_SENDREQUEST_COMPLETE)
+      {
+        // A WinHttpSendRequest API call has completed, complete the current action.
+        CompleteAction();
+      }
+      else if (internetStatus == WINHTTP_CALLBACK_STATUS_WRITE_COMPLETE)
+      {
+        // A WinHttpWriteData call has completed, complete the current action.
+        CompleteAction();
+      }
+      else if (internetStatus == WINHTTP_CALLBACK_STATUS_HEADERS_AVAILABLE)
+      {
+        // Headers for an HTTP response are available, complete the current action.
+        CompleteAction();
+      }
+      else if (internetStatus == WINHTTP_CALLBACK_STATUS_READ_COMPLETE)
+      {
+        // A WinHttpReadData call has completed. Complete the current action, including the amount
+        // of data read.
+        CompleteActionWithData(statusInformationLength);
+      }
+      else if (internetStatus == WINHTTP_CALLBACK_STATUS_HANDLE_CLOSING)
+      {
+        // An HINTERNET handle is closing, complete the outstanding close request.
+        Log::Write(Logger::Level::Verbose, "Closing handle; completing outstanding Close request");
+        CompleteAction();
+      }
     }
```

Comment:
Is it expected that we'd ever fall out of the else without completing the action?  At a minimum we can log the event.

Reply:
We log each event at the start of the `OnHttpStatusOperation` function. But you're right, there is a programming error where you set the expected status but don't actually handle the expected status - it makes sense to log this because it's almost certainly going to be a problem.